client/src/controllers/LoaderController.ts

```diff
@@ -0,0 +1,92 @@
+import { Controller } from '@hotwired/stimulus';
+
+/**
+ * <button
+ *   type="submit"
+ *   class="button button-longrunning"
+ *   data-controller="w-loader"
+ *   data-w-loader-active-class="button-longrunning-active"
+ *   data-w-loader-active-value="{% trans 'Signing in…' %}"
+ *   data-w-loader-duration-seconds-value="40"
+ *   data-action="w-loader#activate"
+ * >
+ *  {% icon name="spinner" %}
+ *  <em data-w-loader-target="label">{% trans 'Sign in' %}</em>
+ * </button>
+ */
+export class LoaderController extends Controller {
+  static classes = ['active'];
+  static targets = ['label'];
+  static values = {
+    active: { type: String, default: 'Loading' },
+    durationSeconds: { type: Number, default: 30 },
+  };
+
+  activeClass: string;
+  activeValue: string;
+  durationSecondsValue: number;
+  labelTarget: HTMLElement;
+  hasLabelTarget: boolean;
+
+  activate() {
+    let originalText;
+    if (this.hasLabelTarget === true) {
+      originalText = this.labelTarget.innerText;
+    }
+    originalText = this.element.textContent;
+
+    this.cancelSpinner(originalText);
+
+    // If client-side validation is active on this form, and is going to block submission of the
+    // form, don't activate the spinner
+    const form = this.element.closest('form');
+
+    if (
```

Comment:
So you mean I should wrap the form and its if statement inside of when the ignoreValidation is true?
Are we doing this because of cases like in the styleguide where the button is not inside a form

Reply:
Yes, that's what I'm thinking - but if there is no usage that would use this except for the style guide it's probably not worth the added complexity.

Reply:
I will check to see if there are more cases like that

Reply:
Great